### PR TITLE
Created a package to save a set of preferences into the device storage

### PR DESCRIPTION
**Pull request by manu0466**
This pull request adds a package that allows to store a set of preferences into device non volatile memory.
The preferences can be saved unencrypted using the `UnencryptedPreferences` struct or encrypted using the `EncryptedPreferences` struct.

This current implementation works on the following platforms:
* Linux
* Windows
* MacOS
* Android
* iOS
* Web 

Since every platform works differently the preferences are stored in different location, the table below shows where are stored the preferences on each platform:
|Platform | Location |
| ------- | ----------------|
| Linux   | $XDG_CONFIG_HOME/{PREFERENCES_APP_DIR} or $HOME/.config/{PREFERENCES_APP_DIR} |
| macOS   | $HOME/Library/Application Support/{PREFERENCES_APP_DIR} |
| Windows |  C:\Users\\$USER\AppData\Roaming\{PREFERENCES_APP_DIR} |
| Android | {PREFERENCES_APP_DIR} |
| iOS     | {PREFERENCES_APP_DIR} |
| Web | LocalStorage |

`PREFERENCES_APP_DIR` is the value provided with the `set_preferences_app_dir`.
**NOTE:** Since in iOS and Android is not possible to know the application data directory the full path must be provided from the user using the `set_preferences_app_dir` function.

### CODE REVIEW

File: packages/crw-preferences/src/io/mod.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            .all(|c| c.is_ascii_alphanumeric() || ['-', '_'].contains(&c))
+}
+
+/// Loads the string representation of a preferences set.
+///
+/// * `name` - key that uniquely identify the preferences set that will be loaded.  
+/// The `name` key can contain only ascii alphanumeric characters or -, _.
+///
+/// # Errors
+/// This function can returns one of the following errors:
```
Comment (bragaz) [inline packages/crw-preferences/src/io/mod.rs:64]:
```suggestion
/// This function returns one of the following errors:
```

File: packages/crw-preferences/src/io/wasm.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    let window = web_sys::window().ok_or(IoError::Unsupported(
+        "global `window` object not found".to_owned(),
+    ))?;
+
+    Ok(Window::local_storage(&window)
+        .map_err(|_| IoError::Unsupported("Local storage not supported".to_owned()))?
+        .ok_or(IoError::Unsupported("Local storage is null".to_owned()))?)
+}
+
+/// Loads the string representation of a preferences set from the browse `LocalStorage`.
```
Comment (bragaz) [inline packages/crw-preferences/src/io/wasm.rs:21]:
```suggestion
/// Loads the string representation of a preferences set from the browser `LocalStorage`.
```

File: packages/crw-preferences/src/io/wasm.rs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    Storage::set_item(&storage, name, value).map_err(|_| IoError::Write)
+}
+
+/// Deletes the data from the browser `LocalStorage`
+pub fn erase(name: &str) {
+    let storage = get_storage();
+
+    if let Ok(storage) = storage {
+        // Make the compiler happy, an error here should neve occur.
```
Comment (bragaz) [inline packages/crw-preferences/src/io/wasm.rs:58]:
```suggestion
        // Make the compiler happy, an error here should never occur.
```

File: packages/crw-preferences/src/io/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        let _ = Storage::set_item(&storage, name, "");
+    }
+}
+
+/// Check if exist a preferences set into the browser `LocalStorage`.
```
Comment (bragaz) [inline packages/crw-preferences/src/io/wasm.rs:63]:
```suggestion
/// Check if there are existent preferences set into the browser `LocalStorage`.
```

File: packages/crw-preferences/src/preferences.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    /// Saves the preferences into the device disk.
+    fn save(&self) -> Result<()>;
+}
+
+/// Deletes a preferences set from the device storage
+pub fn delete(name: &str) {
+    io::erase(name);
+}
+
+/// Checks if exists a preferences set with the provided `name`.
```
Comment (bragaz) [inline packages/crw-preferences/src/preferences.rs:87]:
```suggestion
/// Checks if there are existent preferences set with the provided `name`.
```

**Review by bragaz — COMMENTED** [review]
looks good to me but I think more eyes are needed in order to not loose anything.
Only small docs errs

**Review by RiccardoM — APPROVED** [review]

**Pull request merged**